File: snobot_sim_java/src/main/java/com/snobot/simulator/simulator_components/LazySimDoubleWrapper.java

package com.snobot.simulator.simulator_components;

import edu.wpi.first.hal.SimDouble;
import edu.wpi.first.wpilibj.simulation.SimDeviceSim;
import edu.wpi.first.hal.simulation.SimDeviceDataJNI.SimDeviceInfo;
import org.apache.logging.log4j.Level;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

@SuppressWarnings("PMD.AvoidDuplicateLiterals")
public class LazySimDoubleWrapper
{
    private static final Logger sLOGGER = LogManager.getLogger(LazySimDoubleWrapper.class);

//    private String mDeviceNamfdsfe;
    private final String mDeviceName;
    private final String mValueName;
    private SimDeviceSim mDeviceSim;
    private SimDouble mSimDouble;

    public LazySimDoubleWrapper(String aDeviceName, String aValueName)
    {
        mDeviceName = aDeviceName;
        mValueName = aValueName;
    }

    public double get()
    {
        SimDouble sim = getSimValue();
        if (sim == null)
        {
            sLOGGER.log(Level.ERROR,
                    "Could not get sim value for device '" + mDeviceName + "' (" + mDeviceSim + "), value '" + mValueName + "' (" + mSimDouble + ")");
            return 0;
        }
        else
        {
            return sim.get();
        }
    }

    public void set(double aValue)
    {
        SimDouble sim = getSimValue();
        if (sim == null)
        {
            sLOGGER.log(Level.ERROR,
                    "Could not get sim value for device '" + mDeviceName + "' (" + mDeviceSim + "), value '" + mValueName + "' (" + mSimDouble + ")");
        }
        else
        {
            sim.set(aValue);
        }
    }

    private SimDouble getSimValue()
    {
        if (mSimDouble != null)
        {
            return mSimDouble;
        }

        SimDeviceSim sim = getDeviceSim();
        if (sim != null)
        {
            mSimDouble = sim.getDouble(mValueName);
        }

        return mSimDouble;
    }

    private SimDeviceSim getDeviceSim()
    {
        if (mDeviceSim != null)
        {
            return mDeviceSim;
        }

        SimDeviceInfo[] devices = SimDeviceSim.enumerateDevices(mDeviceName);
        if (devices.length == 1) // NOPMD
        {
            mDeviceSim = new SimDeviceSim(mDeviceName);
        }

        return mDeviceSim;

    }
}

File: snobot_sim/src/main/native/cpp/SnobotSim/SimulatorComponents/LazySimDoubleWrapper.cpp


#include "SnobotSim/SimulatorComponents/LazySimDoubleWrapper.h"
#include "SnobotSim/Logging/SnobotLogger.h"

#include <iostream>

LazySimDoubleWrapper::LazySimDoubleWrapper(
        const std::string& aDeviceName,
        const std::string& aValueName) : mDeviceName(aDeviceName), mValueName(aValueName)
{
    
}


double LazySimDoubleWrapper::get()
{
    setupSimValue();
    if(mSimDouble)
    {
        std::cout << "Getting value..." << mSimDouble << std::endl;
        return mSimDouble.Get();
    }
    
    SNOBOT_LOG(SnobotLogging::LOG_LEVEL_CRITICAL, "Could not get sim value for device '" << mDeviceName << 
        "' (" << mDeviceSimHandle << "), value '" << mValueName << "' (" << mSimDouble << ")"); 
    return 0;
}

void LazySimDoubleWrapper::set(double aValue)
{
    setupSimValue();
    if(mSimDouble)
    {
        std::cout << "Setting value..." << aValue << ", " << mSimDouble << std::endl;
        mSimDouble.Set(aValue);
    }
    else
    {
        SNOBOT_LOG(SnobotLogging::LOG_LEVEL_CRITICAL, "Could not get sim value for device '" << mDeviceName << 
            "' (" << mDeviceSimHandle << "), value '" << mValueName << "' (" << mSimDouble << ")"); 
    }
}


void LazySimDoubleWrapper::setupDeviceSim()
{
    if(mDeviceSimHandle == 0)
    {
        std::stringstream stream;

        stream << "!!!!!!!!!!!!!!Trying to get device... '" << mDeviceName << "' for values '" << mValueName << "'" << std::endl;
        frc::sim::SimDeviceSim::EnumerateDevices(
            mDeviceName.c_str(), [&](const char* name, HAL_SimDeviceHandle handle) {
                stream << "Getting " << name << std::endl;
                if (wpi::StringRef(name) == mDeviceName) {
                    mDeviceSimHandle = handle;
                    stream << "Got it !!!" << std::endl;
                }
            });

        stream << "End value: " << mDeviceSimHandle << std::endl;
        std::cout << stream.str() << std::endl;
    }
}

void LazySimDoubleWrapper::setupSimValue()
{
    if (!mSimDouble)
    {
        setupDeviceSim();
        if (mDeviceSimHandle != 0)
        {
            mSimDouble = HALSIM_GetSimValueHandle(mDeviceSimHandle, mValueName.c_str());
        }
    }
}